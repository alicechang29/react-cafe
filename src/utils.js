const MENU_ATTRIBUTES = ["id", "title", "description", "recipe", "serving"];

/**Normalizes the id string value
 *  formats to lowercase and replaces spaces with dashes  */
function convertToId(id) {
  return id.toLowerCase().replace(" ", "-");
}

export { MENU_ATTRIBUTES, convertToId };