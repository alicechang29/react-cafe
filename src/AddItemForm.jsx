import { useState } from "react";
import { MENU_ATTRIBUTES, convertToId } from "./utils.js";
// import MenuItemInput from "./MenuItemInput.jsx";

/**
 * AddItemForm
 *
 * Props: addItem //FIXME:
 *
 * State: formData //FIXME:
{
  menuItems: {id: '', title: '', description: '', recipe: '', serving: ''}
  type: "snack"
}
 *
 * App -> { AddItemForm } -> MenuItemInput
 */

function AddItemForm({ addItem }) {
  const initialFormData = {
    menuItem: { id: '', name: '', description: '', recipe: '', serve: '' },
    type: "snack"
  };
  console.log("AddItemForm");
  const [formData, setFormData] = useState(initialFormData);

  /** Handle submission of form, normalizes form data inputs, resets form*/
  function handleSubmit(evt) {
    evt.preventDefault();
    console.log("AddItemFormDisplay: handleSubmit", { formData });

    formData.menuItem.id = convertToId(formData.menuItem.id); //TODO: check this

    addItem(formData);

    setFormData(initialFormData);
  }


  /** Update form input. */
  function handleChange(evt) {
    console.log(evt.target);
    const { name, value, option } = evt.target;
    const menuType = document.querySelector("#menu-type").value;

    setFormData(currData => ({
      ...currData,
      menuItem: { ...currData.menuItem, [name]: value },
      type: menuType
    }));

  }

  //TODO: how do i put the label htmlFor for my inputs
  return (
    <form className="AddItemFormDisplay" onSubmit={handleSubmit}>
      <label htmlFor="AddItemFormDisplay-label">Add to the Menu</label>

      {MENU_ATTRIBUTES.map(m => (
        <input
          id={"AddItemFormDisplay" + m}
          key={"AddItemFormDisplay" + m}
          className="form-control"
          value={formData.menuItem[m]}
          name={m}
          placeholder={m}
          required
          onChange={handleChange}
        />
      ))}

      <label htmlFor="menu-type">Snack or Drink?</label>
      <select name="menu-type" id="menu-type" onChange={handleChange}>
        <option value="snacks">Snack</option>
        <option value="drinks">Drink</option>
      </select>


      <button>Add to Menu!</button>
    </form >

  );


}

export default AddItemForm;


/*
{MENU_ATTRIBUTES.map((menuAttr, i) => (
        <MenuItemInput
          key={i}
          menuAttr={menuAttr}
          answer={formData[menuAttr]}
          handleChange={handleChange} />
      ))}



*/