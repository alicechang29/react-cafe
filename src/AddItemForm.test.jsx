import { describe, test as it, expect, vi } from "vitest";
import { screen, fireEvent, render, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import AddItemForm from "./AddItemForm";
import { MENU_ATTRIBUTES } from "./utils";


it("renders without crashing", async function () {
  const { asFragment } = render(
    <MemoryRouter initialEntries={["/add"]}>
      <AddItemForm />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("calls addItem with correct inputs upon save", function () {
  const handleSubmit = vi.fn();
  const { container } = render(
    <MemoryRouter initialEntries={["/add"]}>
      <AddItemForm addItem={handleSubmit} />
    </MemoryRouter>
  );

  fireEvent.change(screen.getByLabelText("name"), {
    target: { value: "test name" },
  });
  fireEvent.change(screen.getByLabelText("description"), {
    target: { value: "test description" },
  });
  fireEvent.change(screen.getByLabelText("recipe"), {
    target: { value: "test recipe" },
  });
  fireEvent.change(screen.getByLabelText("serve"), {
    target: { value: "test serve" },
  });
  const button = container.querySelector("button");

  fireEvent.click(button);
  //FIXME:
  // expect(handleSubmit).toHaveBeenCalled();
  // expect(handleSubmit).toHaveBeenCalledWith("snacks", {
  //   description: "test description",
  //   name: "test name",
  //   recipe: "test recipe",
  //   serve: "test serve",
  // });
});

