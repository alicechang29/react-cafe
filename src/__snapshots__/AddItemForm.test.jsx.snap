// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <form
    class="AddItemForm"
  >
    <label
      for="AddItemForm-label"
    >
      Add to the Menu
    </label>
    <input
      aria-label="id"
      class="form-control"
      id="AddItemForm-id"
      name="id"
      placeholder="id"
      required=""
      value=""
    />
    <input
      aria-label="name"
      class="form-control"
      id="AddItemForm-name"
      name="name"
      placeholder="name"
      required=""
      value=""
    />
    <input
      aria-label="description"
      class="form-control"
      id="AddItemForm-description"
      name="description"
      placeholder="description"
      required=""
      value=""
    />
    <input
      aria-label="recipe"
      class="form-control"
      id="AddItemForm-recipe"
      name="recipe"
      placeholder="recipe"
      required=""
      value=""
    />
    <input
      aria-label="serve"
      class="form-control"
      id="AddItemForm-serve"
      name="serve"
      placeholder="serve"
      required=""
      value=""
    />
    <label
      class="d-inline-flex"
      for="AddItemForm-type"
    >
      Snack or Drink?
    </label>
    <select
      class="form-control form-control-sm d-inline-flex AddItemForm-type"
      id="AddItemForm-type"
      name="type"
    >
      <option
        value="snacks"
      >
        Snack
      </option>
      <option
        value="drinks"
      >
        Drink
      </option>
    </select>
    <button
      class="btn-primary btn btn-sm AddItemForm-addBtn"
    >
      Add to Menu!
    </button>
  </form>
</DocumentFragment>
`;
